### PR TITLE
No more Q states

### PR DESCRIPTION
**Pull request by Raelifin**
I've never seen a "Q-State" outside of Teacher. Q functions take (state,action) pairs. Storing the concatenation of states and actions isn't particularly elegant and only saves us a tiny amount of computation.

Furthermore, concatenating states and actions only really works when they're the same rank. For Atari environments or other non-MuJoCo environments this will not always be the case. We can extend support for more environments by pulling them apart and adding some checks for environment dimensionality.

### CODE REVIEW

**Comment by Raelifin** [comment]
Oh, I also rearrange the main setup logic for Teacher so that it's more likely to immediately crash if something doesn't work, rather than, say, do rollouts and then crash.

**Comment by nottombrown** [comment]
Hmmm, `q_state` is the standard name for the `(obs, action)` pairs that are used in `q-learning`:

https://www.google.com/search?q="q+state"+mdp

However, I think that you're right that keeping the actions and observations explicitly separated until they are fed into the neural net is an improvement. It should make the system more accessible to newcomers.

Have you tested that this PR doesn't cause any performance regressions on Mujoco?

File: rl_teacher/teach.py [hunk 1]
```diff
@@ -54,44 +53,78 @@ def __init__(self, env, summary_writer, comparison_collector, agent_logger, labe
...
+            self.segment_alt_act_placeholder = tf.placeholder(
+                dtype=tf.float32, shape=(None, None) + self.act_shape, name="alt_act_placeholder")
+            # Assume the actions are how we want them
+            segment_act = self.segment_act_placeholder
+            segement_alt_act = self.segment_alt_act_placeholder
+
+        # A vanilla multi-layer perceptron maps a (state, action) pair to a reward (Q-value)
+        mlp = FullyConnectedMLP(self.obs_shape, self.act_shape)
+
+        self.q_value = self._predict_rewards(self.segment_obs_placeholder, segment_act, mlp)
```
Comment (nottombrown) [inline rl_teacher/teach.py]: Good call moving `mlp` to a local variable

File: rl_teacher/teach.py [hunk 2]
```diff
@@ -54,44 +53,78 @@ def __init__(self, env, summary_writer, comparison_collector, agent_logger, labe
...
+            self.segment_act_placeholder = tf.placeholder(
+                dtype=tf.float32, shape=(None, None) + self.act_shape, name="act_placeholder")
+            self.segment_alt_act_placeholder = tf.placeholder(
+                dtype=tf.float32, shape=(None, None) + self.act_shape, name="alt_act_placeholder")
+            # Assume the actions are how we want them
+            segment_act = self.segment_act_placeholder
+            segement_alt_act = self.segment_alt_act_placeholder
+
+        # A vanilla multi-layer perceptron maps a (state, action) pair to a reward (Q-value)
+        mlp = FullyConnectedMLP(self.obs_shape, self.act_shape)
```
Comment (nottombrown) [inline rl_teacher/teach.py]: Good call moving `mlp` to a local variable 👌

Comment (Raelifin) [inline rl_teacher/teach.py]: I haven't tested on MuJoCo, as my license expired and I wasn't able to renew. :(

**Comment by nottombrown** [comment]
Made some small changes and am running regression tests currently

**Comment by nottombrown** [comment]
image

Doesn't break performance (actually improved things on this seed). Merging in!

**Pull request merged**